### PR TITLE
Feature/assignment 7

### PR DESCRIPTION
**Pull request by nikhil1-successive**

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -1,15 +1,30 @@
-import express, { Application } from 'express';
+import express, { Application, Request, Response } from 'express';
 import userRoutes from './routes/index';
 
-const app: Application = express();
-const port: number = 8000;
+class MyApp {
```
Comment (shahbaz-successive) [inline src/app.ts:4]: If class is used then file name should be starting capital letter

File: src/lib/math.ts [hunk 2]
```diff
@@ -1,16 +1,27 @@
-import lodash from "lodash"
+import lodash from "lodash";
 
-function add(num1: number, num2: number): number {
-    return lodash.sum([num1, num2]);
+class MathOperations {
```
Comment (shahbaz-successive) [inline src/lib/math.ts:3]: File name should start with Capital as you're using class function

File: src/middleware/locationMiddleware.ts [hunk 3]
```diff
@@ -1,24 +1,32 @@
 import axios, { AxiosResponse } from "axios";
 import { Request, Response, NextFunction } from "express";
 
-const locationMiddleware = async (req: Request, res: Response, next: NextFunction) => {
-  try {
-    const clientIP: string | undefined = req.ip;
-    const response: AxiosResponse<any, any> = await axios.get(`https://ipinfo.io/${clientIP}/json`);
-    const { country }: any = response.data;
+interface GeoLocationMiddlewareOptions {
```
Comment (shahbaz-successive) [inline src/middleware/locationMiddleware.ts:4]: interface declaration should start with letter I

File: src/utils/mockData.ts [hunk 4]
```diff
@@ -1,12 +1,19 @@
-interface User {
+export interface User {
```
Comment (shahbaz-successive) [inline src/utils/mockData.ts:1]: interface declaration should start with letter I

**Review by shahbaz-successive — CHANGES_REQUESTED** [review]
Please rename all file names where using class function to start with capital letter
Interface declaration should start with capital letter I throughout project

File: src/app.ts (hunk 1 above)
Comment (nikhil1-successive) [reply src/app.ts]: done

File: src/lib/math.ts (hunk 2 above)
Comment (nikhil1-successive) [reply src/lib/math.ts]: done

File: src/middleware/locationMiddleware.ts (hunk 3 above)
Comment (nikhil1-successive) [reply src/middleware/locationMiddleware.ts]: done

File: src/utils/mockData.ts (hunk 4 above)
Comment (nikhil1-successive) [reply src/utils/mockData.ts]: done

File: src/middleware/AuthMiddleware.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    this.authenticate = this.authenticate.bind(this);
+  }
+  //authneticate method
+  authenticate = (req: Request, res: Response, next: NextFunction): void => {
+    const token: string | undefined = req.headers['authorization'];
+    if (!token) {
+      next(createError(403, 'Please provide token'));
+      return;
+    }
+    jwt.verify(token as string, this.secretKey, (err, decodeUser) => {
```
Comment (shahbaz-successive) [inline src/middleware/AuthMiddleware.ts:18]: Provide type to err and decodeUser

File: src/routes/userRoutes.ts [hunk 6]
```diff
@@ -1,32 +1,37 @@
...
+import ParameterValidatorMiddleware from '../middleware/ValidateParamMiddleware';
+import QueryMiddleware from '../middleware/QueryMiddleware';
+import GeoLocationMiddleware from '../middleware/LocationMiddleware'
+import myCustomHeaderMiddleware from '../middleware/customheaderMiddleware';
+import CustomMiddleware from '../middleware/CustomMiddleware';
 
 const router: Router = express.Router();
 const secretKey: string = 'alpha-beta-gamma';
+const rateLimiter = new RateLimiterMiddleware();
+const errorHandler = new ErrorHandlerMiddleware();
```
Comment (shahbaz-successive) [inline src/routes/userRoutes.ts:19]: Add type to all these variables

File: src/routes/userRoutes.ts [hunk 7]
```diff
@@ -1,32 +1,37 @@
...
-// Refer to mockData file for email and password required for authentication
-router.post('/login', customMiddleware, (req: Request, res: Response) => {
+// hit this route by passing email and password in body. Get email and body from mockData.ts file. 
+router.post('/login', (req: Request, res: Response) => {
   const { email, password }: any = req.body;
-  const user: any = userData.find(u => u.email === email && u.password === password);
-  if (user) {
+  const user: IUser | undefined = UserData.getUserByEmail(email);
+  if (user && user.password === password) {
     const token = jwt.sign(
```
Comment (shahbaz-successive) [inline src/routes/userRoutes.ts:35]: Add type to token

File: src/routes/userRoutes.ts [hunk 8]
```diff
@@ -38,33 +43,42 @@ router.post('/login', customMiddleware, (req: Request, res: Response) => {
...
-router.get('/chainmiddleware', customMiddleware, authMiddleware, middleware1, middleware2, (req: Request, res: Response) => {
+// hit this route by passing {Key:'Authorization' and Value:'Token' in Headers}. You will get Token on successfull login
+router.get('/chainmiddleware', authMiddleware.authenticate, (req: Request, res: Response) => {
   res.send("Middleware Called");
 });
 
-router.post('/seeddata', customMiddleware, authMiddleware, (req: Request, res: Response) => {
-  const foodData: string[] = dataSeeder();
+router.post('/seeddata', (req: Request, res: Response) => {
+  const dataSeeder = new DataSeeder();
```
Comment (shahbaz-successive) [inline src/routes/userRoutes.ts:52]: Add type

File: src/utils/RegistrationValidationSchema.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  validate(req: Request, res: Response, next: NextFunction): void {
+    const validationResult = this.userSchema.validate(req.body, { abortEarly: false });
+    if (validationResult.error) {
+      const errors: string[] = validationResult.error.details.map((detail) => detail.message);
+      res.status(400).json({ errors });
+    }
+    next();
+  }
+}
+const registrationValidator = new RegistrationValidator();
```
Comment (shahbaz-successive) [inline src/utils/RegistrationValidationSchema.ts:19]: Add type

File: src/utils/ValidationRules.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  }
+}
+const validationRules: IValidationRules = {
+  login: {
+    email: Joi.string().email().required(),
+    password: Joi.string().min(6).required(),
+    username: Joi.string().alphanum().min(3).max(30).required(),
+  },
+};
+const requestValidator = new RequestValidator(validationRules);
```
Comment (shahbaz-successive) [inline src/utils/ValidationRules.ts:39]: Add type

**Review by shahbaz-successive — CHANGES_REQUESTED** [review]
Feedback done

**Review by shahbaz-successive — APPROVED** [review]
👍

**Pull request merged**